### PR TITLE
📖 document providers conventions for clusterctl

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
clusterctl v2 is designed to work with all the providers.

This PR documents the contract/the conventions each provider should follow in order to work properly with clusterctl. 

NB. More rules will be added while the clusterctl work complete/agreement on open issues is reached. 

**Which issue(s) this PR fixes** :
Rif #https://github.com/kubernetes-sigs/cluster-api/issues/1802

### CODE REVIEW

File: docs/book/src/providers/clusterctl.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# `clusterctl`
+
+## clusterctl v1alpha3 (clusterctl redesign)
+
+`clusterctl` is a CLI tool for handling the lifecycle of a cluster-API management cluster.
+
+The v1alpha3 release is designed for providing a simple a day 1 experience; `clusterctl` is bundled with Cluster API and can be reused across providers
```
Comment (vincepri) [inline docs/book/src/providers/clusterctl.md]:
```suggestion
The v1alpha3 release is designed for providing a simple day 1 experience; `clusterctl` is bundled with Cluster API and can be reused across providers
```

File: docs/book/src/providers/clusterctl.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+### Components YAML
+
+The provider is required to generate a single YAML file with all the components required for installing the provider
+itself (CRD, Controller, RBAC etc.). 
+
+Infrastructure providers MUST release a file called `infrastructure-components.yaml`, while bootstrap provider MUST
+release a file called ` bootstrap-components.yaml` (exception for CABPK, which is included in CAPI by default).
+
+The components YAML should contain one Namespace object only; this namespace will be used as a default target namespace
```
Comment (vincepri) [inline docs/book/src/providers/clusterctl.md]:
```suggestion
The components YAML should contain one Namespace object, which will be used as the default namespace
```

File: docs/book/src/providers/clusterctl.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+for installing the provider components.
```
Comment (vincepri) [inline docs/book/src/providers/clusterctl.md]:
```suggestion
when creating provider components.
```

File: docs/book/src/providers/clusterctl.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+> If the generated component YAML does not contain a Namespace object, `clusterctl` can determine the default namespace;
```
Comment (vincepri) [inline docs/book/src/providers/clusterctl.md]:
```suggestion
> If the generated component YAML doesn't contain a Namespace object, users will need to provide one.
```

File: docs/book/src/providers/clusterctl.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+> in this case the user will be always required to specify a target namespace for installing the provider.
+
+> In case there is more than one Namespace object in the components YAML, `clusterctl` will generate an error and abort
+> the provider installation.
+
+The components YAML can contain environment variables matching the regexp `\${\s*([A-Z0-9_]+)\s*}`; it is highly
+recommended to prefix the variable name with the provider name e.g. `{ $AWS_CREDENTIALS }`
+
+> The users should ensure those variables are set in advance before running `clusterctl`; in case some variable
```
Comment (vincepri) [inline docs/book/src/providers/clusterctl.md]:
```suggestion
> Users are required to ensure that environment variables are set in advance before running `clusterctl`; if a variable
```

File: docs/book/src/providers/clusterctl.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+> are missing at run time, `clusterctl` will generate an error and abort the provider operation.
```
Comment (vincepri) [inline docs/book/src/providers/clusterctl.md]:
```suggestion
> is missing, `clusterctl` will generate an error and abort the provider operation.
```

File: docs/book/src/providers/clusterctl.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+ 
+### Cluster templates
```
Comment (vincepri) [inline docs/book/src/providers/clusterctl.md]:
```suggestion
### Workload cluster templates
```

File: docs/book/src/providers/clusterctl.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+Infrastructure provider could publish cluster templates to be used by `clusterctl config cluster`.
+
+Cluster templates MUST be stored in the same folder of the component YAML and adhere to the the following naming convention:
+1. The default cluster template should be named `config-{bootstrap}.yaml`. e.g `config-kubeadm.yaml`
+2. Additional cluster template should be named `config-{flavor}-{bootstrap}.yaml`. e.g `config-production-kubeadm.yaml`
+
+`{bootstrap}` is the name of the bootstrap provider used in the template; `{flavor}` is the name the user can pass to the
+`clusterctl config cluster --flavor` flag to identify the specific template to use.
+
+## Old versions
```
Comment (vincepri) [inline docs/book/src/providers/clusterctl.md]:
```suggestion
## Previous versions (unsupported) 
```

File: docs/book/src/providers/clusterctl.md (hunk 1 above)
Comment (fabriziopandini) [reply docs/book/src/providers/clusterctl.md]: done

File: docs/book/src/providers/clusterctl.md (hunk 2 above)
Comment (fabriziopandini) [reply docs/book/src/providers/clusterctl.md]: done

File: docs/book/src/providers/clusterctl.md (hunk 3 above)
Comment (fabriziopandini) [reply docs/book/src/providers/clusterctl.md]: done

File: docs/book/src/providers/clusterctl.md (hunk 4 above)
Comment (fabriziopandini) [reply docs/book/src/providers/clusterctl.md]: done

File: docs/book/src/providers/clusterctl.md (hunk 5 above)
Comment (fabriziopandini) [reply docs/book/src/providers/clusterctl.md]: done

File: docs/book/src/providers/clusterctl.md (hunk 6 above)
Comment (fabriziopandini) [reply docs/book/src/providers/clusterctl.md]: done

File: docs/book/src/providers/clusterctl.md (hunk 7 above)
Comment (fabriziopandini) [reply docs/book/src/providers/clusterctl.md]: done

File: docs/book/src/providers/clusterctl.md (hunk 8 above)
Comment (fabriziopandini) [reply docs/book/src/providers/clusterctl.md]: done

**Comment by fabriziopandini** [comment]
@vincepri done!

**Pull request merged**